cirq-core/cirq/transformers/insertion_sort.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer that sorts commuting operations in increasing order of their `.qubits` tuple."""
+
+from typing import Optional, TYPE_CHECKING, List, Tuple, FrozenSet, Union
+
+from cirq import protocols, circuits
+from cirq.transformers import transformer_api
+
+if TYPE_CHECKING:
+    import cirq
+
+_MAX_QUBIT_COUNT_FOR_MASK = 64
+
+
+@transformer_api.transformer(add_deep_support=True)
+def insertion_sort_transformer(
+    circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+) -> 'cirq.Circuit':
+    """Sorts the operations using their `.qubits` property as comparison key.
+
+    Operations are swapped only if they commute.
+
+    Args:
+        circuit: input circuit.
+        context: optional TransformerContext (not used),
+    """
+    all_operations = [*circuit.all_operations()]
+    relative_order = {
+        qs: i for i, qs in enumerate(sorted(set(tuple(sorted(op.qubits)) for op in all_operations)))
+    }
+    if len(circuit.all_qubits()) <= _MAX_QUBIT_COUNT_FOR_MASK:
+        # use bitmasks.
+        q_index = {q: i for i, q in enumerate(circuit.all_qubits())}
+
+        def _msk(qs: Tuple['cirq.Qid', ...]) -> int:
+            msk = 0
+            for q in qs:
+                msk |= 1 << q_index[q]
+            return msk
+
+        operations_with_info: Union[
+            List[Tuple['cirq.Operation', int, int]], List[Tuple['cirq.Operation', int, FrozenSet]]
+        ] = [
```

Comment:
I don't think it is worthwhile to introduce this branching and a duality of type of the third tuple element which can be `int` or `FrozenSet`, both of which are later used with the `&` operator.  The extra human overhead in reading and understanding the code and a higher likelihood of bugs in a more complicated implementation is IMO not worth the time saved on a special case benchmark.

On my host - after upgrading to Python 3.13.0 the original benchmark times are slightly better on your side when it can use bitmask instead of set -
```
# PR 6776 at 9c31fe6f
491 ms ± 21.1 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
593 ms ± 28.1 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
```
However, after increasing the number of qubits to 100, to use the second branch below, the times were
```
# PR 6776 at 9c31fe6f
1.67 s ± 132 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
1.49 s ± 57.3 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
```

To have a more thorough check, I have tried another benchmark `bench_isort_z.ipy`, which needs to check commutation between CZ and Z gates -
```
# bench_isort_z.ipy
import cirq

moment_cz = cirq.Moment(cirq.CZ(cirq.q(i + 1), cirq.q(i)) for i in reversed(range(0, 64, 2)))
moment_z = cirq.Moment(cirq.Z(cirq.q(i)) for i in reversed(range(64)))
circuit = cirq.Circuit(20 * [moment_cz, moment_z])

%timeit -r10 cirq.transformers.insertion_sort_transformer(circuit)
```
For this one, the timings with this PR and the [amend-6776 branch](https://github.com/pavoljuhas/Cirq/commit/ee1e264e47673f5c1255136aacee61a6ef49bf83) show no significant difference and are both likely bound by the `protocols.commutes` calls:
```
# PR 6776 at 9c31fe6f
4.04 s ± 166 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
3.99 s ± 96.6 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
```